ui/server: add ability to create conditional statement diagnostics

lindseyjin opened the pull request:
Resolves #57634

Previously, we did not have the ability to create conditional statement
diagnostics in the frontend. This commit adds in the ability to specify
a minimum execution latency and an expiry time when creating a statement
diagnostics request. These changes apply to both DB and CC console.

Since expired requests are not surfaced at all in the frontend, we have
also modified the statement diagnostics API response to not return
already expired and incomplete requests.

Lastly, this commit also deletes some unused code related to statement
diagnostics modals.

Release note (ui change): Add ability to create conditional statement
diagnostics by adding two new fields 1) minimum execution latency, which
specifies the limit for when a statement should be tracked and 2) expiry
time, which specifies when a diagnostics request should expire.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/74112)
<!-- Reviewable:end -->

lindseyjin committed:
ui/server: add ability to create conditional statement diagnostics

Resolves #57634

Previously, we did not have the ability to create conditional statement
diagnostics in the frontend. This commit adds in the ability to specify
a minimum execution latency and an expiry time when creating a statement
diagnostics request. These changes apply to both DB and CC console.

Since expired requests are not surfaced at all in the frontend, we have
also modified the statement diagnostics API response to not return
already expired and incomplete requests.

Lastly, this commit also deletes some unused code related to statement
diagnostics modals.

Release note (ui change): Add ability to create conditional statement
diagnostics by adding two new fields 1) minimum execution latency, which
specifies the limit for when a statement should be tracked and 2) expiry
time, which specifies when a diagnostics request should expire.

lindseyjin reviewed (commented):
CC Console:
https://www.loom.com/share/32056d2af2f4429a805bd07b1dbf11f6

DB Console:
https://www.loom.com/share/c053b2ae5bec42f58610aac2ecf40bc2


@annebirzin Let me know if the styling looks good to you! 
I think there's a few minor details that are off eg) I'm missing the alert left blue border, because I couldn't find a way to override the alert styling on both CC and DB console. Hope that's still ok, but let me know if you think otherwise!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74112#-:-MrYF6Ps40-B8hb_XPTJ:b1btmdd)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/server/statement_diagnostics_requests.go, line 184 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74112#-MrYGC273OXlA_Q65pX7:-MrYGC273OXlA_Q65pX8:b-g9heyi) ([raw file](https://github.com/cockroachdb/cockroach/blob/679a73005d8f0e99b577b499d488a52da2376f8a/pkg/server/statement_diagnostics_requests.go#L184)):*
> ```Go
> 				req.ExpiresAt = expiresAt.Time
> 				// Don't return already expired requests.
> 				if req.ExpiresAt.Before(time.Now()) {
> ```

I decided to modify the backend API response, since we're currently not surfacing expired requests at all in the frontend.

If we have a use for them in the future (e.g. showing cancelled/expired requests), we can probably include them back in the response and add a selector in the frontend instead. 
___
*[pkg/ui/workspaces/cluster-ui/src/statementsDiagnostics/activateStatementDiagnosticsModal.tsx, line 167 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74112#-MrYGg9rBsiWervE3SNr:-MrYGg9s-fvicV6tZUW-:bqz2mqi) ([raw file](https://github.com/cockroachdb/cockroach/blob/679a73005d8f0e99b577b499d488a52da2376f8a/pkg/ui/workspaces/cluster-ui/src/statementsDiagnostics/activateStatementDiagnosticsModal.tsx#L167)):*
> ```XML
>                     </Select>
>                   </div>
>                   <Divider type="horizontal" style={{ marginBottom: 0 }} />
> ```

Inline styling, since including a class on the antd divider won't override default styling
___
*[pkg/ui/workspaces/cluster-ui/src/statementsDiagnostics/activateStatementDiagnosticsModal.scss, line 4 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74112#-MrYHBOT1H110f4iqIM0:-MrYHBOU6djyw_DvfjUs:bi78eks) ([raw file](https://github.com/cockroachdb/cockroach/blob/679a73005d8f0e99b577b499d488a52da2376f8a/pkg/ui/workspaces/cluster-ui/src/statementsDiagnostics/activateStatementDiagnosticsModal.scss#L4)):*
> ```SCSS
> 
> .modal-body {
>   .ant-modal-body{
> ```

CC console has some global styling rules that overrides [certain styling](https://github.com/cockroachlabs/managed-service/blob/a2ae91082ff2803270fc1c6006177a4dd08d5a2a/console/assets/css/global/_overrides.scss#L11) rules, but that doesn't exist on DB console. Not sure if it's good practice to override antd class names like this, but don't think there's a better way around this?


<!-- Sent from Reviewable.io -->

lindseyjin reviewed (commented):
Modal with updated description text:

![image.png](https://files.reviewableusercontent.io/images/reviewable/29153209/Yzmx_9CLDVUD3zXkKepDJqdt/image.png)

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74112#-:-MrYKJCLAOuh8HEa4nU8:bll05rd)* status: :shipit: complete! 0 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

maryliag reviewed (commented):
> Reviewed 19 of 24 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74112)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @lindseyjin and @maryliag)

___
*[-- commits, line 15 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/74112#-MrYLBkxEA_ZjkbNGaF5):*
nit: add the issue number 
___
*[pkg/ui/workspaces/cluster-ui/src/store/statementDiagnostics/statementDiagnostics.sagas.spec.ts, line 33 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/74112#-MrYNzav9YfJKzibu5AQ:-MrYNzav9YfJKzibu5AR:bruer1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c12550e3c207bb4fcbd4fd5a8e94b8816e52d99/pkg/ui/workspaces/cluster-ui/src/store/statementDiagnostics/statementDiagnostics.sagas.spec.ts#L33)):*
> ```XML
> 
> describe("statementsDiagnostics sagas", () => {
>   describe("createDiagnosticsReportSaga", () => {
> ```

can you also add new tests checking if after the expire date the request is no longer there?


<!-- Sent from Reviewable.io -->

jocrl reviewed (changes requested):
Woo! Commented a few things; the main one is to avoid using refs and imperative patterns in react. Happy to chat if it would help 🙂

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74112#-:-MrYWz4r7kaaaUy5BPYL:b6oatpc)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jocrl, @lindseyjin, and @maryliag)

___
*[pkg/ui/workspaces/cluster-ui/src/statementDetails/statementDetails.tsx, line 74 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/74112#-MrYVwTrCV6F-xmfThFp:-MrYVwTrCV6F-xmfThFq:b-fysigs) ([raw file](https://github.com/cockroachdb/cockroach/blob/3996db66bb3c2cc730c9efaba2c03082f5d433ab/pkg/ui/workspaces/cluster-ui/src/statementDetails/statementDetails.tsx#L74)):*
> ```XML
>   ActivateStatementDiagnosticsModal,
> } from "../statementsDiagnostics";
> type IDuration = google.protobuf.IDuration;
> ```

I'm kinda suspicious about the introduction and use of this `google.protobuf.IDuration` type here. It seems like it wasn't present in the codebase prior to this PR, and I feel like the codebase surely has encountered some sort of this same "duration" type before.

I see in `status.proto` that `min_execution_latency` has the type `google.protobuf.Duration`. Searching for that type in `.proto` files, another field with this type is `aggregation_interval`, which in the front end`aggregationInterval` uses type `number`.

Would that work for this?
___
*[pkg/ui/workspaces/cluster-ui/src/statementDetails/statementDetails.tsx, line 334 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/74112#-MrYN1EbDCG9FzTfkkcN:-MrYN1EbDCG9FzTfkkcO:b-hl2nsv) ([raw file](https://github.com/cockroachdb/cockroach/blob/3996db66bb3c2cc730c9efaba2c03082f5d433ab/pkg/ui/workspaces/cluster-ui/src/statementDetails/statementDetails.tsx#L334)):*
> ```XML
>   StatementDetailsState
> > {
>   activateDiagnosticsRef: React.RefObject<ActivateDiagnosticsModalRef>;
> ```

In the other comment I mention that I think using a ref here is a bad idea. But if you were to keep the ref, the name here of the interface `ActivateDiagnosticsModalRef` is kinda weird. As your property name `activateDiagnosticsRef` implies, the entire thing of type `React.RefObject<ActivateDiagnosticsModalRef>;` is a ref.
However, the point of the `RefObject<T>` generic typing is that `activateDiagnosticsRef` is a ref of an object with the type `T`. That is, `T` itself is not a ref, it is what the ref is pointing to.

I might suggest the name `ActivateDiagnosticsModal`, but that's already used for the component 😛. Maybe `ActivateDiagnosticsModalElement`, in keeping with the `HTMLElement` type name ((code example)[https://stackoverflow.com/questions/61102101/cannot-assign-refobjecthtmldivelement-to-refobjecthtmlelement-instance]).



_Code quote:_
```XML
ActivateDiagnosticsModalRef
```
___
*[pkg/ui/workspaces/cluster-ui/src/statementDetails/statementDetails.tsx, line 346 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/74112#-MrYPPlC4Iyvuc-RIK_V:-MrYPPlC4Iyvuc-RIK_W:b-wrko7n) ([raw file](https://github.com/cockroachdb/cockroach/blob/3996db66bb3c2cc730c9efaba2c03082f5d433ab/pkg/ui/workspaces/cluster-ui/src/statementDetails/statementDetails.tsx#L346)):*
@lindseyjin, is there a particular reason you're "connecting/linking" the `<ActivateStatementDiagnosticsModal/>` and `<DiagnosticsView/>` components with a ref, instead of local state on `<StatementDetails/>`?

I'm curious what other people think here, but I'm inclined to say that the modal should be opened not by calling `showModalFor` on the ref, but by passing as props to `<DiagnosticsView/>` a function that sets local state on `<StatementDetails/>` for whether or not the modal is visible. In `<ActivateStatementDiagnosticsModal/>`, instead of the `useImperativeHandle` that sets state, you would have a `useEffect`.

Refs in general are an anti-pattern in React, and in general there should be a deliberate reason of solving something you cannot otherwise do if you're going to use one. See [these docs](https://pl.reactjs.org/docs/refs-and-the-dom.html), and to quote:

> Your first inclination may be to use refs to “make things happen” in your app. If this is the case, take a moment and think more critically about where state should be owned in the component hierarchy. Often, it becomes clear that the proper place to “own” that state is at a higher level in the hierarchy. See the Lifting State Up guide for examples of this.

In general with react, you want data and props to just "flow", instead of "telling" another component to do something. It's telling that the docs for a ref start with warnings to not use it 😉

In this case it seems like only these two components are concerned with whether the modal is open, so local state is a good idea. If more components were concerned, that might be a reason to hold that state in redux.
___
*[pkg/ui/workspaces/cluster-ui/src/statementsDiagnostics/activateStatementDiagnosticsModal.scss, line 4 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74112#-MrYHBOT1H110f4iqIM0:-MrYUead9eiPmtiA57Fp:b-x24qyu) ([raw file](https://github.com/cockroachdb/cockroach/blob/679a73005d8f0e99b577b499d488a52da2376f8a/pkg/ui/workspaces/cluster-ui/src/statementsDiagnostics/activateStatementDiagnosticsModal.scss#L4)):*
<details><summary><i>Previously, lindseyjin (Lindsey Jin) wrote…</i></summary><blockquote>

CC console has some global styling rules that overrides [certain styling](https://github.com/cockroachlabs/managed-service/blob/a2ae91082ff2803270fc1c6006177a4dd08d5a2a/console/assets/css/global/_overrides.scss#L11) rules, but that doesn't exist on DB console. Not sure if it's good practice to override antd class names like this, but don't think there's a better way around this?
</blockquote></details>

Yeah, if I'm understanding that `ant-modal-body` is a class from the 3rd party library, I do agree this is kinda weird 😛. I'm assuming the problem is that just applying `modal-body` doesn't take precedence over the padding specified in the library on `modal-body`?

I see in your other comment that you apply an inline style to the ant Divider. Could you do something similar, and instead of passing `className`, pass inline styles to style the modal? Since presumably other use cases that want to style the modal would also run into this problem.


<!-- Sent from Reviewable.io -->

maryliag reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74112)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jocrl, @lindseyjin, and @maryliag)

___
*[pkg/ui/workspaces/cluster-ui/src/statementsDiagnostics/activateStatementDiagnosticsModal.scss, line 4 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74112#-MrYHBOT1H110f4iqIM0:-MrY_iSx8O9gqNB6tSAh:b-1ck0o8) ([raw file](https://github.com/cockroachdb/cockroach/blob/679a73005d8f0e99b577b499d488a52da2376f8a/pkg/ui/workspaces/cluster-ui/src/statementsDiagnostics/activateStatementDiagnosticsModal.scss#L4)):*
<details><summary><i>Previously, jocrl (Josephine) wrote…</i></summary><blockquote>

Yeah, if I'm understanding that `ant-modal-body` is a class from the 3rd party library, I do agree this is kinda weird 😛. I'm assuming the problem is that just applying `modal-body` doesn't take precedence over the padding specified in the library on `modal-body`?

I see in your other comment that you apply an inline style to the ant Divider. Could you do something similar, and instead of passing `className`, pass inline styles to style the modal? Since presumably other use cases that want to style the modal would also run into this problem.
</blockquote></details>

I prefer not using inline styles because is hard to keep track and modify all at once. We do have some other instances where we change the `ant` styles, like here: https://github.com/cockroachdb/cockroach/blob/b8323204a3e13d0ff49b2234f6fa2b8907f5fb4e/pkg/ui/workspaces/db-console/src/views/app/components/modal/styles.styl#L19
so I think is okay to override the rules here


<!-- Sent from Reviewable.io -->

jocrl reviewed (commented):
Also, just wanted to add a congratulations on the number of moving parts for this feature 😁!

And I can't speak for anyone else's expectations, but given that you're also leaving tomorrow on my end I would 10/10 understand if you left things for Thomas to deal with 😉. Which would also be a good learning experience for him! I think the first section of https://pl.reactjs.org/docs/refs-and-the-dom.html (up to and including "Don’t Overuse Refs") is worth really reading and understanding, if you plan to write more react in the future.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74112#-:-MrY_n6vC9xCrLU1f5KR:b-c1eyy6)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jocrl, @lindseyjin, and @maryliag)



<!-- Sent from Reviewable.io -->

maryliag reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74112)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jocrl, @lindseyjin, and @maryliag)

___
*[pkg/server/statement_diagnostics_requests.go, line 184 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74112#-MrYGC273OXlA_Q65pX7:-MrYaQMsDk470KzIqK0D:bem2bw9) ([raw file](https://github.com/cockroachdb/cockroach/blob/679a73005d8f0e99b577b499d488a52da2376f8a/pkg/server/statement_diagnostics_requests.go#L184)):*
<details><summary><i>Previously, lindseyjin (Lindsey Jin) wrote…</i></summary><blockquote>

I decided to modify the backend API response, since we're currently not surfacing expired requests at all in the frontend.

If we have a use for them in the future (e.g. showing cancelled/expired requests), we can probably include them back in the response and add a selector in the frontend instead. 
</blockquote></details>

Make sense to update the code here :smile: 
Just one thing, is that you will get a lint error using `time` here (error: `forbidden; use 'timeutil' instead`) 


<!-- Sent from Reviewable.io -->

jocrl reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74112)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jocrl, @lindseyjin, and @maryliag)

___
*[pkg/ui/workspaces/cluster-ui/src/statementsDiagnostics/activateStatementDiagnosticsModal.scss, line 4 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74112#-MrYHBOT1H110f4iqIM0:-MrYaeC73Yaaap6g6Ccn:b44tpx) ([raw file](https://github.com/cockroachdb/cockroach/blob/679a73005d8f0e99b577b499d488a52da2376f8a/pkg/ui/workspaces/cluster-ui/src/statementsDiagnostics/activateStatementDiagnosticsModal.scss#L4)):*
<details><summary><i>Previously, maryliag (Marylia Gutierrez) wrote…</i></summary><blockquote>

I prefer not using inline styles because is hard to keep track and modify all at once. We do have some other instances where we change the `ant` styles, like here: https://github.com/cockroachdb/cockroach/blob/b8323204a3e13d0ff49b2234f6fa2b8907f5fb4e/pkg/ui/workspaces/db-console/src/views/app/components/modal/styles.styl#L19
so I think is okay to override the rules here
</blockquote></details>

👍


<!-- Sent from Reviewable.io -->

jocrl reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74112)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jocrl, @lindseyjin, and @maryliag)

___
*[pkg/ui/workspaces/cluster-ui/src/statementDetails/statementDetails.tsx, line 346 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/74112#-MrYPPlC4Iyvuc-RIK_V:-MrYbrKsEUs-wc6YJ4A0:b-amz7j2) ([raw file](https://github.com/cockroachdb/cockroach/blob/3996db66bb3c2cc730c9efaba2c03082f5d433ab/pkg/ui/workspaces/cluster-ui/src/statementDetails/statementDetails.tsx#L346)):*
Actually, one edit to my suggestion:

Above, I suggested
> instead of the `useImperativeHandle` that sets state, you would have a `useEffect`.

I think instead of a `useEffect`, you want to lift the `visible` state out of `<ActivateStatementDiagnosticsModal/>`, and hold it as a single source of truth in `<DiagnosticsView/>`, as opposed to duplicating the `visible` state in both components. So you would just pass `setVisible` as props to `<ActivateStatementDiagnosticsModal/>`.

The companion docs to "don't use refs" is [lifting state up](https://reactjs.org/docs/lifting-state-up.html) (which, the docs sadly are still using class components). I think this docs page is also pretty core react that's worth taking time to understand 🙂.


<!-- Sent from Reviewable.io -->

Annebirzin commented:
@lindseyjin Looking great! A couple of updates:

- [ ] I chatted with @kevin-v-ngo today about making it so the blue alert message only shows when the 2nd collection method is selected and the expiration date is deselected (so the blue message would not show if the first collection method is selected and the expiration is deselected - I've added that state to [Figma here](https://www.figma.com/file/xdmwvnFQd6KkO9RJ0XLDH0/22.1_SQL-obsrv_query-performance?node-id=5178%3A142577) for reference. 
- [ ] In the Loom videos it looked like the blue used in the controls and links was off. Can we use our core blue (#0055FF) for radio buttons, check boxes and links? Also the background box for the alert message blue looked a little off (the stroke and the fill) 

Let me know any questions

jocrl reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74112)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jocrl, @lindseyjin, and @maryliag)

___
*[pkg/ui/workspaces/cluster-ui/src/statementsDiagnostics/activateStatementDiagnosticsModal.tsx, line 49 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/74112#-MrYt88N-XZmfUS32moC:-MrYt88O01kMvLrog8_p:bm9pzwk) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c12550e3c207bb4fcbd4fd5a8e94b8816e52d99/pkg/ui/workspaces/cluster-ui/src/statementsDiagnostics/activateStatementDiagnosticsModal.tsx#L49)):*
> ```XML
>     ref: React.RefObject<ActivateDiagnosticsModalRef>,
>   ) => {
>     const [visible, setVisible] = useState(false);
> ```

This state is the one that I think would be a good candidate to lift up, now that the parent component also needs to use it 🙂
___
*[pkg/ui/workspaces/cluster-ui/src/statementsDiagnostics/activateStatementDiagnosticsModal.tsx, line 109 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/74112#-MrYsxV-3BFdXLhWqzHs:-MrYsxV02GcM3e6d9Dkg:bz9onsp) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c12550e3c207bb4fcbd4fd5a8e94b8816e52d99/pkg/ui/workspaces/cluster-ui/src/statementsDiagnostics/activateStatementDiagnosticsModal.tsx#L109)):*
> ```XML
>     return (
>       <Modal
>         visible={visible}
> ```

Here it looks like you are passing `visible` to some reusable `Modal` component, that presumably will follow whatever value gets passed for this?


<!-- Sent from Reviewable.io -->

jocrl reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74112)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jocrl, @lindseyjin, and @maryliag)

___
*[pkg/ui/workspaces/cluster-ui/src/statementDetails/statementDetails.tsx, line 334 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/74112#-MrYN1EbDCG9FzTfkkcN:-MrZ2ALN9Blv0FfpfE-r:bv9obt8) ([raw file](https://github.com/cockroachdb/cockroach/blob/3996db66bb3c2cc730c9efaba2c03082f5d433ab/pkg/ui/workspaces/cluster-ui/src/statementDetails/statementDetails.tsx#L334)):*
<details><summary><i>Previously, jocrl (Josephine) wrote…</i></summary><blockquote>

In the other comment I mention that I think using a ref here is a bad idea. But if you were to keep the ref, the name here of the interface `ActivateDiagnosticsModalRef` is kinda weird. As your property name `activateDiagnosticsRef` implies, the entire thing of type `React.RefObject<ActivateDiagnosticsModalRef>;` is a ref.
However, the point of the `RefObject<T>` generic typing is that `activateDiagnosticsRef` is a ref of an object with the type `T`. That is, `T` itself is not a ref, it is what the ref is pointing to.

I might suggest the name `ActivateDiagnosticsModal`, but that's already used for the component 😛. Maybe `ActivateDiagnosticsModalElement`, in keeping with the `HTMLElement` type name ((code example)[https://stackoverflow.com/questions/61102101/cannot-assign-refobjecthtmldivelement-to-refobjecthtmlelement-instance]).
</blockquote></details>

Retracting since there is another usage that uses the same naming.
___
*[pkg/ui/workspaces/cluster-ui/src/statementDetails/statementDetails.tsx, line 346 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/74112#-MrYPPlC4Iyvuc-RIK_V:-MrZ1z1h8KHEXEpct-XU:bbs7h20) ([raw file](https://github.com/cockroachdb/cockroach/blob/3996db66bb3c2cc730c9efaba2c03082f5d433ab/pkg/ui/workspaces/cluster-ui/src/statementDetails/statementDetails.tsx#L346)):*
<details><summary><i>Previously, jocrl (Josephine) wrote…</i></summary><blockquote>

Actually, one edit to my suggestion:

Above, I suggested
> instead of the `useImperativeHandle` that sets state, you would have a `useEffect`.

I think instead of a `useEffect`, you want to lift the `visible` state out of `<ActivateStatementDiagnosticsModal/>`, and hold it as a single source of truth in `<DiagnosticsView/>`, as opposed to duplicating the `visible` state in both components. So you would just pass `setVisible` as props to `<ActivateStatementDiagnosticsModal/>`.

The companion docs to "don't use refs" is [lifting state up](https://reactjs.org/docs/lifting-state-up.html) (which, the docs sadly are still using class components). I think this docs page is also pretty core react that's worth taking time to understand 🙂.
</blockquote></details>

Lindsey brought up that this imperative pattern already existed and was not introduced in this PR. She made this issue https://github.com/cockroachdb/cockroach/issues/74221 to refactor out the refs.


<!-- Sent from Reviewable.io -->

lindseyjin reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74112)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jocrl and @maryliag)

___
*[pkg/ui/workspaces/cluster-ui/src/statementDetails/statementDetails.tsx, line 74 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/74112#-MrYVwTrCV6F-xmfThFp:-MrZGRWI0Yk5V7VvJlH9:bmiq52d) ([raw file](https://github.com/cockroachdb/cockroach/blob/3996db66bb3c2cc730c9efaba2c03082f5d433ab/pkg/ui/workspaces/cluster-ui/src/statementDetails/statementDetails.tsx#L74)):*
<details><summary><i>Previously, jocrl (Josephine) wrote…</i></summary><blockquote>

I'm kinda suspicious about the introduction and use of this `google.protobuf.IDuration` type here. It seems like it wasn't present in the codebase prior to this PR, and I feel like the codebase surely has encountered some sort of this same "duration" type before.

I see in `status.proto` that `min_execution_latency` has the type `google.protobuf.Duration`. Searching for that type in `.proto` files, another field with this type is `aggregation_interval`, which in the front end`aggregationInterval` uses type `number`.

Would that work for this?
</blockquote></details>

I know that for `aggregationInterval`, it's passed in as a Duration but we do convert it to a number later on. 

However, `minExecLatency` and `expiresAfter` need to be converted from numbers to Durations to be passed to the backend. I could do this at a point closer to the API request and pass around numbers for the most part, if you think that would be cleaner? However, not sure if there's a huge difference since we'll have to convert it at some point anyways.
___
*[pkg/ui/workspaces/cluster-ui/src/statementDetails/statementDetails.tsx, line 346 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/74112#-MrYPPlC4Iyvuc-RIK_V:-MrZI01HBVnT9bDklOFo:bnrkrle) ([raw file](https://github.com/cockroachdb/cockroach/blob/3996db66bb3c2cc730c9efaba2c03082f5d433ab/pkg/ui/workspaces/cluster-ui/src/statementDetails/statementDetails.tsx#L346)):*
<details><summary><i>Previously, jocrl (Josephine) wrote…</i></summary><blockquote>

Lindsey brought up that this imperative pattern already existed and was not introduced in this PR. She made this issue https://github.com/cockroachdb/cockroach/issues/74221 to refactor out the refs.
</blockquote></details>

Thanks for linking the issue!
___
*[pkg/ui/workspaces/cluster-ui/src/statementsDiagnostics/activateStatementDiagnosticsModal.tsx, line 49 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/74112#-MrYt88N-XZmfUS32moC:-MrZIQw8ExlctTjYZCYX:bhomqmw) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c12550e3c207bb4fcbd4fd5a8e94b8816e52d99/pkg/ui/workspaces/cluster-ui/src/statementsDiagnostics/activateStatementDiagnosticsModal.tsx#L49)):*
<details><summary><i>Previously, jocrl (Josephine) wrote…</i></summary><blockquote>

This state is the one that I think would be a good candidate to lift up, now that the parent component also needs to use it 🙂
</blockquote></details>

hi! wondering if this comment still applies, or if this is part of the followup issue to refactor out refs?

if so, how do I lift up the state? 🤔
___
*[pkg/ui/workspaces/cluster-ui/src/statementsDiagnostics/activateStatementDiagnosticsModal.tsx, line 109 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/74112#-MrYsxV-3BFdXLhWqzHs:-MrZIkF-5teaWdw0ejGj:b-oatxhq) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c12550e3c207bb4fcbd4fd5a8e94b8816e52d99/pkg/ui/workspaces/cluster-ui/src/statementsDiagnostics/activateStatementDiagnosticsModal.tsx#L109)):*
<details><summary><i>Previously, jocrl (Josephine) wrote…</i></summary><blockquote>

Here it looks like you are passing `visible` to some reusable `Modal` component, that presumably will follow whatever value gets passed for this?
</blockquote></details>

From what I understand, we call the `showModalFor` function on the ref from other components in order to set `visible` to true, which allows the modal to show up 
___
*[pkg/ui/workspaces/cluster-ui/src/store/statementDiagnostics/statementDiagnostics.sagas.spec.ts, line 33 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/74112#-MrYNzav9YfJKzibu5AQ:-MrZQmEc13X6Gho-CtDd:bh0dmua) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c12550e3c207bb4fcbd4fd5a8e94b8816e52d99/pkg/ui/workspaces/cluster-ui/src/store/statementDiagnostics/statementDiagnostics.sagas.spec.ts#L33)):*
<details><summary><i>Previously, maryliag (Marylia Gutierrez) wrote…</i></summary><blockquote>

can you also add new tests checking if after the expire date the request is no longer there?
</blockquote></details>

Actually, there isn't any logic in the frontend right now to check if reports are expired, since I've only modified the backend API response to not return expired reports.

I've created a quick test in the backend to check that expired reports shouldn't be returned in the API response instead!


<!-- Sent from Reviewable.io -->

lindseyjin reviewed (commented):
Updated the alert to only show when the second collection method is selected without an expiry date

I've also updated the radio buttons to be the core blue colour. Otherwise, I checked the code and the links/checkbox are already core blue (might've just rendered strangely on my monitor?) :thinking: 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74112#-:-MrZZIaqAZzdTRfu99wD:b3yjxpm)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jocrl and @maryliag)



<!-- Sent from Reviewable.io -->

maryliag reviewed (commented):
Can you just update the images to show the new versions after the changes requested from Anne?

> Reviewed 2 of 24 files at r1, 2 of 4 files at r4, 3 of 3 files at r5, 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74112#-:-Mrc4gBzCK91zGoT3v0g:b4n7yov)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jocrl and @maryliag)

___
*[pkg/ui/workspaces/cluster-ui/src/statementDetails/statementDetails.tsx, line 74 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/74112#-MrYVwTrCV6F-xmfThFp:-Mrc-Rf-3wk8_02zPh46:b-ruu42i) ([raw file](https://github.com/cockroachdb/cockroach/blob/3996db66bb3c2cc730c9efaba2c03082f5d433ab/pkg/ui/workspaces/cluster-ui/src/statementDetails/statementDetails.tsx#L74)):*
<details><summary><i>Previously, lindseyjin (Lindsey Jin) wrote…</i></summary><blockquote>

I know that for `aggregationInterval`, it's passed in as a Duration but we do convert it to a number later on. 

However, `minExecLatency` and `expiresAfter` need to be converted from numbers to Durations to be passed to the backend. I could do this at a point closer to the API request and pass around numbers for the most part, if you think that would be cleaner? However, not sure if there's a huge difference since we'll have to convert it at some point anyways.
</blockquote></details>

I was able to find some usage of IDuration on the codebase, so you can keep like this :) 
___
*[pkg/ui/workspaces/cluster-ui/src/statementsDiagnostics/activateStatementDiagnosticsModal.tsx, line 49 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/74112#-MrYt88N-XZmfUS32moC:-Mrc1uXt61cBYP8Dmwxy:b-isqnk6) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c12550e3c207bb4fcbd4fd5a8e94b8816e52d99/pkg/ui/workspaces/cluster-ui/src/statementsDiagnostics/activateStatementDiagnosticsModal.tsx#L49)):*
<details><summary><i>Previously, lindseyjin (Lindsey Jin) wrote…</i></summary><blockquote>

hi! wondering if this comment still applies, or if this is part of the followup issue to refactor out refs?

if so, how do I lift up the state? 🤔
</blockquote></details>

This is not related to the refactor of refs, but I think we can keep it like this, and work on improvements when working on the refs refactor


<!-- Sent from Reviewable.io -->

lindseyjin reviewed (commented):
New Loom recording with the new changes:
https://www.loom.com/share/afb17abb48a2400aa9cdd69fc8363552

Screenshot:
![image.png](https://files.reviewableusercontent.io/images/reviewable/29153209/D9qNTk9r9r9OVshAlELxHqvt/image.png)

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74112#-:-MrcCzug1EsKjm1UeQWN:by3l9tm)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jocrl and @maryliag)

___
*[pkg/ui/workspaces/cluster-ui/src/statementDetails/statementDetails.tsx, line 74 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/74112#-MrYVwTrCV6F-xmfThFp:-MrcCyhr1nxm5BGL1Wwj:bcr5s2o) ([raw file](https://github.com/cockroachdb/cockroach/blob/3996db66bb3c2cc730c9efaba2c03082f5d433ab/pkg/ui/workspaces/cluster-ui/src/statementDetails/statementDetails.tsx#L74)):*
<details><summary><i>Previously, maryliag (Marylia Gutierrez) wrote…</i></summary><blockquote>

I was able to find some usage of IDuration on the codebase, so you can keep like this :) 
</blockquote></details>

Got it, thanks!


<!-- Sent from Reviewable.io -->

lindseyjin reviewed (commented):
Updated again with slightly different text: s/duration thresholds/latency threshold

![image.png](https://files.reviewableusercontent.io/images/reviewable/29153209/t_E5Ry4mn67nZBlxH-A-6JBH/image.png)

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74112#-:-MrcQ4_z4DkLh3eGeTyU:b3o073p)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jocrl, @maryliag, and @THardy98)



<!-- Sent from Reviewable.io -->

lindseyjin reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74112)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jocrl, @maryliag, and @THardy98)

___
*[pkg/ui/workspaces/cluster-ui/src/util/convert.ts, line 14 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/74112#-MrcWLIS0fw4bBNYPbhZ:-MrcWLIS0fw4bBNYPbh_:b-xtxmfk) ([raw file](https://github.com/cockroachdb/cockroach/blob/6c1b6114dc6f309bbf6083c1240cfb0c9c8ea4ee/pkg/ui/workspaces/cluster-ui/src/util/convert.ts#L14)):*
> ```XML
> 
> import * as protos from "@cockroachlabs/crdb-protobuf-client";
> import { fromNumber } from "long";
> ```

Workaround because there are some files in `db_console` that call functions in this file, and importing `Long` directly introduces errors in the pipeline since the `Long` versions are incompatible


<!-- Sent from Reviewable.io -->

maryliag reviewed (approved):
great work! :tada: 
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 2 of 2 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74112#-:-MrcoL9J-JHV5L2XCD_T:b-s6uj6c)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @jocrl and @THardy98)



<!-- Sent from Reviewable.io -->

lindseyjin commented:
bors r+
